### DISCUSSION TITLE
[Enhancement] Replace Xamarin.Forms.Xaml with System.Xaml

### DISCUSSION THREAD

**Discussion by legistek** [dotnet/maui#85] [General] [answered]
Wondering if anyone's considered dropping in System.Xaml from dotnet/wpf to replace Xamarin.Forms.Xaml

Having contributed to Xamarin.Forms.Xaml a couple of times now, I think it's fair to say it has a lot of technical debt, magic strings, etc., left over from the earliest days of XF that it might be time to consider retiring it. If it's possible to get both WPF and MAUI using the same XAML parser I think that would potentially solve a lot of current limitations - particularly when it comes to naming and namespaces - and make maintenance a lot easier.

I believe, if I'm not mistaken, either the WPF team made sure to make System.Xaml not have any WPF dependencies, or this was already the case. 

Just something to consider if y'all aren't already.

**Comment by memsranga** [comment]
This makes sense. One thing is System.Xaml has a lot more capabilities (like MultiBinding) that Xamarin.Core.Xaml doesn't currently support.

**Comment by davidbritch** [comment]
@shanranm MultiBinding has been added to Xamarin.Forms by @legistek , and will drop in 4.7.

**Comment by legistek** [comment]
Indeed! :) 

Actually @shanranm `Binding`, `MultBinding`, etc., are in `PresentationFramework` not `System.Xaml`. That's a testament to the great job the WPF team did in separating the concerns by making `System.Xaml` flexible enough to support something like data binding while being completely agnostic to how it's implemented.

**Comment by legistek** [comment]
Just to give an example of what I'm talking about:

https://github.com/dotnet/maui/blob/build/System.Maui.Xaml/XamlParser.cs#L105

That sortof stuff's got to go.

**Comment by mrwcjoughin** [comment]
this is exactly why I ditched XAML a long time time. I build all my Xamarin.Forms UI's using code, and will be taking that further soon by using Comet.

thankfully MVU is baked into MAUI so hopefully it won't be too hard to change the comet code to use MAUI's equivalent in November 2021

**Comment by legistek** [comment]
I was looking at the `System.Xaml` code in the WPF repo and I might have spoken a little too soon about the dependencies. It looks like there might be some circular or dummy dependencies on things like `DependencyProperty`, which makes sense because you can't directly assign a `BindingBase` object to a POCO property. 
So it might not work as a drop-in replacement out of the box but probably would need a few modifications.

**Comment by PureWeen** [comment]
@StephaneDelcroix

**Comment by StephaneDelcroix** [comment, marked as answer]
System.Xaml isn't available on any platform we have to support (think netstandard). Also, the exposed API isn't enough to get us a fast, performant, and complete Xaml parser/inflater.

It is in our plan to decouple the Xaml parser from XF code, ideally while moving to MAUI, but I kind of disagree to:

> I think it's fair to say it has a lot of technical debt, magic strings
